### PR TITLE
Fee pooling SDK change

### PR DESCRIPTION
**Pull request by shiqizng**
This PR is to support creating transactions with a fee under mintxnfee since other txns in the group will be paying.

### CODE REVIEW

File: src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java [hunk 1]
```diff
@@ -24,6 +32,7 @@
     public Transaction applicationTransaction = null;
     public SignedTransaction signedTransaction = null;
 
+
```
Comment (jdtzmn) [inline src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java:35]: nit: extra line

File: src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java [hunk 2]
```diff
@@ -99,9 +108,9 @@ public void buildApplicationTransactions(String operation, Long applicationId, S
         if (StringUtils.isNotEmpty(appAccounts)) {
             builder.accounts(convertAccounts(appAccounts));
         }
-        if (fee != 0) {
-            builder.flatFee(fee);
-        }
+
+        builder.flatFee(fee);
```
Comment (jdtzmn) [inline src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java:112]: Not a huge deal, but I would move this up to line 93 so the property setters and conditionals are each grouped together.

File: src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java [hunk 3]
```diff
@@ -120,9 +129,30 @@ public void sign_the_transaction() throws NoSuchAlgorithmException {
...
+        assertThat(txn).containsKey("fee");
+    }
+
+    @Then("fee field not in txn")
+    public void  fee_field_not_in_txn() throws IOException {
+        if (signedTransaction == null) return;
+
+        byte[] encodedTxn = Encoder.encodeToMsgPack(signedTransaction);
+        Map<String,Object> sigtxn = Encoder.decodeFromMsgPack(encodedTxn,Map.class);
+        assertThat(sigtxn).isNotNull();
```
Comment (jdtzmn) [inline src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java:148]: Why is this check not in the `fee_field_is_in_txn()` method?

File: src/test/java/com/algorand/algosdk/unit/Rekeying.java [hunk 4]
```diff
@@ -22,12 +27,17 @@
...
+    Address pk;
+    Transaction txn;
+    MultisigAddress msig;
 
     @Given("payment transaction parameters {int} {int} {int} {string} {string} {string} {int} {string} {string}")	
     public void payment_transaction_parameters(
             Integer fee, Integer fv, Integer lv, String gh,
             String to, String close, Integer amt, String gen, String note) {
 
+
```
Comment (jdtzmn) [inline src/test/java/com/algorand/algosdk/unit/Rekeying.java:40]: nit: extra line

File: src/test/java/com/algorand/algosdk/unit/Rekeying.java [hunk 5]
```diff
@@ -70,4 +80,31 @@ public void the_signed_transaction_should_equal_the_golden(String golden) throws
     public void i_set_the_from_address_to(String fromAddress) {
         transactionBuilder.sender(fromAddress);
     }
+
+    @Given("multisig addresses {string}")
```
Comment (jdtzmn) [inline src/test/java/com/algorand/algosdk/unit/Rekeying.java:84]: Are these related to fee pooling? Or are they just sneaking into this PR for convenience?

**Review by jdtzmn — CHANGES_REQUESTED** [review]

File: src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java (hunk 3 above)
Comment (shiqizng) [reply src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java:148]: Added.

File: src/test/java/com/algorand/algosdk/unit/Rekeying.java (hunk 5 above)
Comment (shiqizng) [reply src/test/java/com/algorand/algosdk/unit/Rekeying.java:83]: these are steps for fee pooling tests, https://github.com/algorand/algorand-sdk-testing/blob/master/features/unit/feetest.feature#L29-L31. I had to put these in the Rekeying class because that's where the payment transaction builder is implemented. Cucumber doesn't allow me to duplicate the same definition in a different file.

File: src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java [hunk 6]
```diff
@@ -82,6 +90,8 @@ public void buildApplicationTransactions(String operation, Long applicationId, S
         }
 
         builder.sender(sender);
+        builder.flatFee(fee);
+
```
Comment (jdtzmn) [inline src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java:94]:
```suggestion
```

nit: extra line

File: src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java [hunk 7]
```diff
@@ -99,9 +109,7 @@ public void buildApplicationTransactions(String operation, Long applicationId, S
         if (StringUtils.isNotEmpty(appAccounts)) {
             builder.accounts(convertAccounts(appAccounts));
         }
-        if (fee != 0) {
-            builder.flatFee(fee);
-        }
+
```
Comment (jdtzmn) [inline src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java:112]:
```suggestion
```
nit: extra line

File: src/test/java/com/algorand/algosdk/unit/Rekeying.java (hunk 5 above)
Comment (jdtzmn) [reply src/test/java/com/algorand/algosdk/unit/Rekeying.java:83]: Gotcha - that makes sense.

**Review by jdtzmn — APPROVED** [review]
2 extra lines but otherwise looks good!

File: src/test/java/com/algorand/algosdk/transaction/TestTransaction.java [hunk 8]
```diff
@@ -81,6 +83,7 @@ public void testPaymentTransactionJsonSerialization() throws Exception {
                 .firstValid(301)
                 .lastValid(1300)
                 .genesisHash(new Digest())
+                .fee(0)
```
Comment (jasonpaulos) [inline src/test/java/com/algorand/algosdk/transaction/TestTransaction.java:86]: What happens if this line is not included? In order to not have a breaking change, we'll want the behavior for existing code like this to remain the same.

File: src/main/java/com/algorand/algosdk/builder/transaction/TransactionBuilder.java [hunk 9]
```diff
@@ -77,13 +77,14 @@ final public Transaction build() {
             } catch (NoSuchAlgorithmException e) {
                 throw new UnsupportedOperationException(e);
             }
+            if (txn.fee == null || txn.fee == BigInteger.valueOf(0)) {
```
[the remark this answers is not in the document]
Comment (jasonpaulos) [reply src/main/java/com/algorand/algosdk/builder/transaction/TransactionBuilder.java:80]: @winder could you weigh in?

[the remark this answers is not in the document]
Comment (shiqizng) [reply src/main/java/com/algorand/algosdk/builder/transaction/TransactionBuilder.java:80]: I think == is right since txn.fee is an int and BigInteger.valueOf(0) is also an int.

[the remark this answers is not in the document]
Comment (jasonpaulos) [reply src/main/java/com/algorand/algosdk/builder/transaction/TransactionBuilder.java:80]: I think `BigInteger.valueOf(0)` returns a BigInteger though

[the remark this answers is not in the document]
Comment (shiqizng) [reply src/main/java/com/algorand/algosdk/builder/transaction/TransactionBuilder.java:80]: you're right.  txn.fee is also a BigInteger object. I'll update == to equals().

File: src/test/java/com/algorand/algosdk/transaction/TestTransaction.java (hunk 8 above)
Comment (shiqizng) [reply src/test/java/com/algorand/algosdk/transaction/TestTransaction.java]: I updated the default fee in Transaction to 0 so it's consistent with the default in TransactionBuilder and removed this line.

File: src/test/java/com/algorand/algosdk/transaction/TestTransaction.java [hunk 10]
```diff
@@ -766,4 +768,101 @@ public void EmptyByteArraysShouldBeRejected() throws Exception {
...
+                    .assetAmount(amountToSend)
+                    .fee(0)
+                    .firstValid(firstValidRound)
+                    .lastValid(lastValidRound)
+                    .genesisHash(gh)
+                    .assetIndex(assetIndex)
+                    .build();
+            assertThat(tx.fee).isEqualTo(1000);
+        }
+    }
```
Comment (jasonpaulos) [inline src/test/java/com/algorand/algosdk/transaction/TestTransaction.java:866]: Could you add more one test to verify that the default behavior when not calling `fee` or `flatFee` results in a transaction with a fee value of 1000? (That is the expected behavior right?)

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java [hunk 11]
```diff
@@ -32,7 +32,7 @@
     @JsonProperty("snd")
     public Address sender = new Address();
     @JsonProperty("fee")
-    public BigInteger fee = Account.MIN_TX_FEE_UALGOS;
+    public BigInteger fee = BigInteger.valueOf(0);
```
Comment (jasonpaulos) [inline src/main/java/com/algorand/algosdk/transaction/Transaction.java:35]: nit: you could use `BigInteger.ZERO` instead of this constant.

**Review by jasonpaulos — CHANGES_REQUESTED** [review]
Mostly good! I just want to verify that the default behavior when no fee is specified is the same. I'll approve after a test for that is added.

File: src/test/java/com/algorand/algosdk/transaction/TestTransaction.java (hunk 10 above)
Comment (shiqizng) [reply src/test/java/com/algorand/algosdk/transaction/TestTransaction.java:866]: I actually changed default fee to 0. I'll switch it back to 1000.

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java (hunk 11 above)
Comment (winder) [reply src/main/java/com/algorand/algosdk/transaction/Transaction.java:35]: Change the default constructor to `public Transaction() {}`, it no longer needs to override the fee after this change.

File: src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java (hunk 6 above)
Comment (winder) [reply src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java:94]: One concern here is that in cucumber, there is no "not set" value so we've always used 0 instead. So in theory this could change the intent of some tests.

I suppose if all the tests pass we weren't using that feature for this field

File: src/test/java/com/algorand/algosdk/transaction/TestTransaction.java [hunk 12]
```diff
@@ -766,4 +768,127 @@ public void EmptyByteArraysShouldBeRejected() throws Exception {
         assertThat(tx.note).isNull();
         assertThat(tx.lease).isNull();
     }
+
+    @Nested
```
Comment (winder) [inline src/test/java/com/algorand/algosdk/transaction/TestTransaction.java:772]: This is the first time I've never seen this feature, nice way to group similar tests! :+1:

**Review by jasonpaulos — APPROVED** [review]
Pending tests, looks good
